src/other/index.jl

```diff
@@ -46,14 +46,36 @@ function names!(x::Index, nms::Vector{Symbol}; makeunique::Bool=false)
 end
 
 function rename!(x::Index, nms)
+    xbackup = copy(x)
+    toholder = Dict()
     for (from, to) in nms
         from == to && continue # No change, nothing to do
-        if haskey(x, to)
+        if !haskey(xbackup, from)
+            copy!(x.lookup, xbackup.lookup)
+            x.names .= xbackup.names
+            error("Tried renaming $from to $to, when $from does not exist in the Index.")
+        end
+        if haskey(toholder, to)
+            copy!(x.lookup, xbackup.lookup)
+            x.names .= xbackup.names
             error("Tried renaming $from to $to, when $to already exists in the Index.")
         end
-        x.lookup[to] = col = pop!(x.lookup, from)
+        if haskey(x, to)
+            toholder[to] = x.lookup[to]
+        end
+        if !haskey(toholder, from) && !haskey(x.lookup, from)
+            copy!(x.lookup, xbackup.lookup)
+            x.names .= xbackup.names
+            error("Tried renaming $from to multiple values.")
+        end
+        x.lookup[to] = col = haskey(toholder, from) ? pop!(toholder, from) : pop!(x.lookup, from)
```

Comment:
```suggestion
        col = pop!(toholder, from, pop!(x.lookup, from))
        x.lookup[to] = col
```

We try to avoid doing multiple assignments in one line. Also I have shortened the expression a bit (but I am OK to leave it as you implemented it).

Reply:
Interesting, the default argument is not lazy:

```julia
julia> a = Dict(:B=>1,:C=>3)
Dict{Symbol,Int64} with 2 entries:
  :B => 1
  :C => 3

julia> b = Dict(:B=>2)
Dict{Symbol,Int64} with 1 entry:
  :B => 2

julia> pop!(a, :B, pop!(b, :B))
1

julia> a
Dict{Symbol,Int64} with 1 entry:
  :C => 3

julia> b
Dict{Symbol,Int64} with 0 entries
```

Reply:
You are right. This is my mistake.